search/utils.py

```diff
@@ -0,0 +1,150 @@
+"""
+Utility module for the search engine, contains various functions to do common
+operations on lists and set of iterables
+"""
+import re
+from search import config
+
+
+def _dec(fl):
+    """Return a stringified more readable float, for debugging purposes."""
+    return '{:.2f}'.format(fl)
+
+
+def normalize(iterable):
+    """
+    Normalize an iterable of numbers in a series that sums up to 1.
+
+    Arguments:
+        iterable (numbers): an iterable containing numbers (either `int` or
+            `float`) to normalize.
+
+    Returns:
+        list: Normalized float values, in the same order as the one provided.
+
+    Example:
+        >>> normalize([3, 2, 1])
+        [0.5, 0.333333, 0.166667]
+
+    """
+    tot = sum(iterable)
+    return [float(v) / tot for v in iterable]
+
+
+def scale_to_one(iterable):
+    """
+    Scale an iterable of numbers proportionally such as the highest number
+    equals to 1
+
+    Example:
+        >>> scale_to_one([5,4,3,2,1])
+        [1, 0.8, 0.6, 0.4, 0.2]
+    """
+    m = max(iterable)
+    return [v / m for v in iterable]
+
+
+def weighted_avg(values, weights):
+    """Calculate the weighted mean average between two iterables of `values`
+    and matching `weights`
+
+    Args:
+        values(iterable): Values to average, either `int` or `float`
+        weights(iterable): Matching weights iterable
+
+    Returns:
+        float: weighted average
+
+    """
+    if len(values) != len(weights):
+        raise ValueError(
+            'Values and Weights length do not match for weighted average')
+    val = sum(m * w for m, w in zip(values, weights))
+    weights = sum(weights)
+    return val / weights
+
+
+def tokenize(string):
+    """
+    Given a string return a list of segments of the string,
+    splitted with :any:`config.STR_SPLIT_REGEX`, removing every word <=
+    :any:`config.MIN_WORD_LENGTH`.
+
+    Example:
+        >>> utils.tokenize('hello there, how are you')
+        ['hello', 'there']
+
+    """
+    return [
+        w for w in re.split(config.STR_SPLIT_REGEX, string)
+        if len(w) > config.MIN_WORD_LENGTH
+    ]
+
+
+def get_max_moves(list_, idx):
```

Comment:
Questa è la distanza dal bordo della lista prendendo il massimo dei due. Quindi è:
```max_distance```
E però facciamo capire che stiamo parlando di parole e di frasi. ```list_``` è in realtà ```phrase```, ```ìdx``` è ```word_index```, etc...